src/pal/src/misc/environ.cpp

```diff
@@ -634,11 +660,424 @@ SetEnvironmentVariableA(
+
+    CPalThread * pthrCurrent = InternalGetCurrentThread();
+    InternalEnterCriticalSection(pthrCurrent, &gcsEnvironment);
+
+    for (int i = 0; palEnvironment[i] != nullptr; ++i)
+    {
+        const char *equalsSignPosition = strchr(palEnvironment[i], '=');
+        if (equalsSignPosition == nullptr)
+        {
+            equalsSignPosition = palEnvironment[i] + strlen(palEnvironment[i]);
+        }
+
+        // Check whether the name of this variable has the same length as the one
+        // we're looking for before proceeding to compare them.
+        if (equalsSignPosition - palEnvironment[i] == nameLength)
+        {
+            if (memcmp(name, palEnvironment[i], nameLength) == 0)
+            {
+                // Free the string we're removing.
+                InternalFree(palEnvironment[i]);
+
+                // Move the last environment variable pointer here.
+                palEnvironment[i] = palEnvironment[palEnvironmentCount - 1];
+                palEnvironment[palEnvironmentCount - 1] = nullptr;
+
+                palEnvironmentCount--;
+            }
+        }
+    }
+
+    InternalLeaveCriticalSection(pthrCurrent, &gcsEnvironment);
+}
+
+/*++
+Function:
+  EnvironPutenv
+
+Add the environment variable string provided to the PAL version
+of the environment.
+
+Parameters
+
+    entry
+            [in] The variable string to add. Should be in the format
+                 "name=value", where value might be empty (see below).
+    deleteIfEmpty
+            [in] If this is TRUE, "name=" will unset the 'name' variable.
+
+Return Values
+
+    TRUE on success, FALSE otherwise
+
+--*/
+BOOL EnvironPutenv(const char* entry, BOOL deleteIfEmpty)
+{
+    bool fOwningCS = false;
+    BOOL result = FALSE;
+
+    CPalThread * pthrCurrent = InternalGetCurrentThread();
+
+    const char *equalsSignPosition = strchr(entry, '=');
+    if (equalsSignPosition == entry || equalsSignPosition == nullptr)
+    {
+        // "=foo" and "foo" have no meaning
+        goto done;
+    }
+
+    if (equalsSignPosition[1] == '\0' && deleteIfEmpty)
+    {
+        // "foo=" removes foo from the environment in _putenv() on Windows.
+        // The same string can result from a call to SetEnvironmentVariable()
+        // with the empty string as the value, but in that case we want to
+        // set the variable's value to "". deleteIfEmpty will be FALSE in
+        // that case.
+        int length = strlen(entry);
+        char* copy = (char *) InternalMalloc(length);
+        if (copy == nullptr)
+        {
+            goto done;
+        }
+
+        memcpy(copy, entry, length - 1);
+
+        // Change '=' to '\0'
+        copy[length - 1] = '\0';
+
+        EnvironUnsetenv(copy);
+        InternalFree(copy);
+
+        result = TRUE;
+    }
+    else
+    {
+        // See if we are replacing an item or adding one.
+
+        // Make our copy up front, since we'll use it either way.
+        char* copy = strdup(entry);
```

Comment:
I would prefer the strdup. You can then move the strdup out of the if / else. 
When I was looking at it now, I have also noticed that we can get rid of the strlen in the if branch. If you move the nameLength assignment before the if, then you can use it instead of `length - 1` in the `copy[length - 1] = '\0'`. Would you mind making this change as well?